Wrappers/Python/cil/optimisation/functions/SVRGFunction.py

```diff
@@ -0,0 +1,260 @@
+    
+        .. math ::
+            x_{k+1} = x_k - \gamma [n*\nabla f_i(x_k) - n*\nabla f_i(\tilde{x}) + \nabla \sum_{i=0}^{n-1}f_i(\tilde{x})],    where :math:`\tilde{x}` is the latest "snapshot" point . Note that compared with the literature, we multiply by :math:`n`, the number of functions, so that we return an approximate gradient of the whole sum function and not an average gradient. 
+    
+    Reference: Johnson, R. and Zhang, T., 2013. Accelerating stochastic gradient descent using predictive variance reduction. Advances in neural information processing systems, 26.
+    
+
+    Parameters
+    ----------
+     functions : `list`  of functions
+        A list of functions: :code:`[f_{0}, f_{1}, ..., f_{n-1}]`. Each function is assumed to be smooth with an implemented :func:`~Function.gradient` method. All functions must have the same domain. The number of functions must be strictly greater than 1. 
+    sampler: An instance of a CIL Sampler class ( :meth:`~optimisation.utilities.sampler`) or of another class which has a `next` function implemented to output integers in {0,...,n-1}.
+        This sampler is called each time gradient is called and  sets the internal `function_num` passed to the `approximate_gradient` function.  Default is `Sampler.random_with_replacement(len(functions))`. 
+    update_frequency : int or None, optional
+        The frequency of updating the full gradient (taking a snapshot). The default is 2*len(functions) so a "snapshot" is taken every 2*len(functions) iterations. 
+    store_gradients : bool, default: `False`
+        Flag indicating whether to store an update a list of gradients for each function :math:`f_i` or just to store the snapshot point :math:` \tilde{x}` and its gradient :math:`\nabla \sum_{i=0}^{n-1}f_i(\tilde{x})`.
+
+    
+    """
+
+    def __init__(self, functions, sampler=None, update_frequency=None, store_gradients=False):
+        super(SVRGFunction, self).__init__(functions, sampler)
+
+        # update_frequency for SVRG
+        self.update_frequency = update_frequency
+    
+        if self.update_frequency is None:
+            self.update_frequency = 2*self.num_functions
+        self.store_gradients = store_gradients
+
+        self._svrg_iter_number = 0
+        
+        self._full_gradient_at_snapshot = None
+        if self.store_gradients:
+            self._list_stored_gradients = None
+        
+        self.snapshot = None 
+
+    def gradient(self, x, out=None):
+        """ Selects a random function using the `sampler` and then calls the approximate gradient at :code:`x` or calculates a full gradient depending on the update frequency
+
+        Parameters
+        ----------
+        x : DataContainer (e.g. ImageData object)
```

Comment:
Nice.  I didn't see ImageData mentioned in the SAG/A one so please add there as well.

Reply:
DOne